### PR TITLE
Block ngram repeats

### PR DESCRIPTION
**Pull request by KaijuML**
Hi guys,

Tagging @pltrdy following a previous conversation.

In this pull request I fix the accuracy issue of block_ngram_repeats (see issue #1445), as well as improve on the efficiency of the method.

In the current implementation, there are two unintended behaviors:
    - when all beams contain forbidden repeats, one of them is chosen arbitrarily and the translations fails from there because it is invalid, resulting in a generation truncated at this point. (As an other negative side effect, the beam search process still goes through its min_length iterations even-though nothing happens from there, each time trying to select a valid beam and failing).
    - when not all beams contain forbidden repeats, we allow the beam search to start only from the valid ones, significantly reducing the value of the beam search (we "wasted" beams by allowing them to go in unwanted direction just to trash them at the next iteration)

This is not the desired behaviors: we want a mechanism that prevents beams from even containing forbidden repeats, and block them ahead of time. 
In this pull request, we look at which tokens would result in forbidden repeats and simply put their log-probas at -1e20, preventing the beam from going in this direction, but keeping the beam alive if any other direction is good.

Furthermore I also improve on the efficiency of the method by keeping track step by step of the generated tokens so that the search for forbidden_tokens is just a dictionary lookup, whereas it previously was a whole for-loop.

Apart from rewriting the block_ngram_repeats method I introduce minimal changes:
   - in __init__ of decode strategy when create self.forbidden_tokens for the dictionary lookup
   - in beam.advance I moved the self.block_ngram_repeats(self, logprobs) to directly before the next paths are selected
    - in beam.advance I add a call to self.maybe_update_forbidden_tokens when the next paths have been selected
     - I wrote the maybe_update_forbidden_tokens in decode_strategy

Let me know if we need anything more, and if everything works as intended

### CODE REVIEW

**Comment by pltrdy** [comment]
@BigKaiju thx for the PR. I'll test it today :)

 I openned a PR on your repo to fix this one. Just merge it and should pass the CI tests here then.

**Comment by pltrdy** [comment]
@BigKaiju do you have results before/after this PR? 

I've seen lower rouge scores using the PR, around 1 ROUGE point on 2 different models (and different data)

**Comment by BigKaiju** [comment]
I have seen improvement one my task, in BLEU score. This is also due to the fact that when generating long sequence (500+ words) with beam_size=5, it usually happens that all 5 beams contain repeats, leading to catastrophic failure (issue #1445): the generated sequence will stop immediately, in mid-sentence, in mid-document. I have noticed that for sequences that were un-interrupted, results were different with my proposed implementation from the existing implementation; it must be due to the fact that now, the generation always work with 5 beams, whereas before when one beam was censored (due to repeats) it would be lost.

I don't see any error in the code. The only error prone part is where I re-order the dictionnaries, if you want to take an extra careful look overthere. It's in `decode_strategy.py` at method `maybe_update_forbidden_tokens`

**Comment by vince62s** [comment]
@BigKaiju We have tried this on a very simple translation task, where there is no repetition of tokens.
This PR gives us a lower BLEU and different output whereas it should give us the exact same result, rigth ?

**Comment by BigKaiju** [comment]
@vince62s Are you referring to the unittests `onmt.tests.test_beam_search.py` that are currently failing ? I have inspected them and found two blocking points:
 - When we iterate over `i`, we check if length is long enough to trigger repeats with `i <= ngram_repeat ` (see  [line 50](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/tests/test_beam_search.py#L50)); however with the proposed implementation this should be with a strict inequality `i < ngram_repeat`, because as I said we block repeats before advancing any beam in a forbidden direction (whereas current implementation advances non-the-less and after drops the beam). If we make this change in the unittest, we pass the first set of assertTrue (see [line 54](https://github.com/OpenNMT/OpenNMT-py/blob/master/onmt/tests/test_beam_search.py#L54)). The second issue is more of an issue with the way the test is written: when finding a repeat, the current implementation blocks the whole beam: all scores go to -10e20. However, recall that the proposed implementation works one step ahead, only changing the scores that would result in a forbidden beam. This means that when we push scores with -inf everywhere but a 0 at repeat_idx, words resulting in forbidden beams are at -10e20 but all other stay at -inf (contrary to current implementation that pushes everything to -10e20). This results in final scores that still contain -inf whereas the unittest check for equality with -10e20. I don't have time right now to pinpoint exactly why this happens but I think this is not problematic. We would have to rewrite the test somehow I we decide to use the proposed implementation.

**Comment by BigKaiju** [comment]
@vince62s @pltrdy Please also note that I have just fixed an issue where a dict was passed as a ref instead of value, leading to shared info between beams. It now works as intended, and yields better results at the task I am working on.

**Comment by pltrdy** [comment]
@BigKaiju thanks for your updates, I'll be testing it soon!

File: onmt/translate/decode_strategy.py [hunk 1]
```diff
@@ -118,6 +122,77 @@ def block_ngram_repeats(self, log_probs):
...
+
+        n = self.block_ngram_repeat
+
+        forbidden_tokens = list()
+        for path_idx, seq in zip(self.select_indices, self.alive_seq):
+
+            # Reordering forbidden_tokens following beam selection
+            # We rebuild a dict to ensure we get the value and not the pointer
+            forbidden_tokens.append(
+                {k:v for k,v in self.forbidden_tokens[path_idx]})
```
Comment (pltrdy) [inline onmt/translate/decode_strategy.py]: I guess it should be:
```
{k:v for k,v in self.forbidden_tokens[path_idx].items()})
```
or even better:
```
forbidden_tokens.append(dict(self.forbidden_tokens[path_idx]))
```

Comment (BigKaiju) [reply onmt/translate/decode_strategy.py]: Yes of course, my bad. Which ever you prefer!

**Comment by vince62s** [comment]
@BigKaiju Can you please fix Flake and the unittest ? it would be great to merge before the codebase changes too much. Thanks.

**Comment by pltrdy** [comment]
@BigKaiju I've a local version which pass the test

**Comment by BigKaiju** [comment]
> @BigKaiju I've a local version which pass the test

Do you mean you already modified the tests, or did you change the code ?

**Comment by pltrdy** [comment]
@BigKaiju Well I've been struggling a bit with git here, finally doing a clean rebase on another branch/PR.

Thanks a lot for your contributions @BigKaiju :)

**Pull request closed without merging**